api/src/main/java/org/pmiops/workbench/notebooks/LeonardoApiClientFactory.java

```diff
@@ -0,0 +1,61 @@
+package org.pmiops.workbench.notebooks;
+
+import com.google.auth.oauth2.OAuth2Credentials;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Provider;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.firecloud.FirecloudApiClientFactory;
+import org.pmiops.workbench.leonardo.ApiClient;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class LeonardoApiClientFactory {
+  private final FirecloudApiClientFactory firecloudApiClientFactory;
+  private final Provider<WorkbenchConfig> workbenchConfigProvider;
+
+  @Autowired
+  public LeonardoApiClientFactory(
+      FirecloudApiClientFactory firecloudApiClientFactory,
+      Provider<WorkbenchConfig> workbenchConfigProvider) {
+    this.firecloudApiClientFactory = firecloudApiClientFactory;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+  }
+
+  public ApiClient newImpersonatedApiClient(String userEmail) throws IOException {
+    OAuth2Credentials delegatedCreds =
+        firecloudApiClientFactory.getDelegatedUserCredentials(userEmail);
+    ApiClient client = newApiClient();
```

Comment:
Optional: Initializing ApiClient is a heavy load operation and encourage to reuse.
But in this case, I am not sure what is the best way to reuse because we need two different types of API clients. One way can be two static API Client, one is regular, another one is for Impersonated?

Reply:
First - why do you say it's a heavy load operation? Is it because of the `OkHttpClient` creation?

There is only one created per offline task in my egress engine PR. I would not want to reuse API clients across two different impersonated users or hold any static state, this seems unnecessarily risky to me. There should still be the same number of APIClients before and after this PR. I added some comments here to clarify how this should be used. There are up to 3 types for each APIClient:

1. authenticated as RW SA
2. authenticated as the end user (during a request)
3. authenticated as an impersonated user

The first two will continue to be cached for the duration of the request scope only (if used) - at most 1 instance, via NotebooksConfig, FirecloudConfig.

Reply:
Ah, I did not realize we uses OkHttpClient. Previously I use Jersey and its javadoc
https://docs.oracle.com/javaee/7/api/javax/ws/rs/client/Client.html mentioned:
`Clients are heavy-weight objects that manage the client-side communication infrastructure. `
Since then I have been very careful of reusing it. But since this is the existing behaviour,  we can keep it as it.